pkg/config/context.go

```diff
@@ -16,7 +17,11 @@ func SyncContextsAndServers() error {
 		return errors.Wrap(err, "failed to get client config")
 	}
 
-	config.PopulateContexts(cfg)
+	hasChanged := config.PopulateContexts(cfg)
+	needsSync := doServersNeedUpdate(cfg)
+	if !hasChanged && !needsSync {
+		return nil
+	}
 
 	// Now write the context to the configuration file.  This will also create any missing server for its corresponding context
 	for _, c := range cfg.KnownContexts {
```

Comment:
This optimization is quite clean and much appreciated.

I am not sure if we had in the past rely on this to update existing (not create new) entry to correct for any inconsistencies. cc @prkalle 
If we want to be really safe, we can consider some undocumented env var to restore back the old behavior for any user who needs it.

Reply:
Looking at the #616 and what was being done before, I see that we were calling https://github.com/vmware-tanzu/tanzu-plugin-runtime/blob/7278b2554915a10ad3abc9e8befe0a3eb196e1a3/config/legacy_clientconfig_factory.go#L21.  I based my check on that code to know if there was any servers needing to be updated.  So I feel we are covered on that front.

However, there is another aspect to consider.  When saving the configuration to file I notice that there is some cleanup of "legacyDiscardedContexts" done here: https://github.com/vmware-tanzu/tanzu-plugin-runtime/blob/7278b2554915a10ad3abc9e8befe0a3eb196e1a3/config/config_factory.go#L144-L152

To reproduce the situation I installed `cluster:k8s:v0.25.4` and simply ran `tanzu cluster version`.  What I see is that the `config.yaml` file now has a `contexts` node with all the servers included there.  Before this PR, that `contexts` node would be removed on the next CLI run.  With this PR it is only cleaned up whenever the configuration is changed later on.

I don't think keeping this `contexts` node is problematic, but let's discuss it

I

Reply:
We have revisited the solution and are now much more thorough.